src/binary.ts

```diff
@@ -0,0 +1,259 @@
+
+    let version = buf.version();
+    if (!(version instanceof Error)) {
+      if (!isManaged && semver.lt(version, minVersion)) {
+        return new Error(`Detected installed Buf CLI ${version}, which does not support LSP.\nSee https://buf.build/docs/installation for installation instructions.`);
+      } else if (isManaged) {
+        // Run this asynchronously. This avoids blocking startup on doing an an
+        // HTTP GET to GitHub's API.
+        this.checkForUpdates(version).then((result) => {
+          if (result instanceof Error) {
+            vscode.window.showErrorMessage(`${result}`);
+          }
+        });
+      }
+      return buf;
+    }
+
+    // Switch over to managed mode, because we couldn't find an installed `buf`.
+    if (!isManaged) {
+      buf.path = this.bundledPath();
+    }
+
+    // Need to block here.
+    let result = await this.checkForUpdates();
+    if (result instanceof Error) {
+      return new Error(`Failed to update installed Buf CLI: ${result}`);
+    }
+    return buf;
+  }
+
+  async checkForUpdates(current?: string): Promise<Result<void>> {    
+    let release = await github.release('bufbuild', 'buf', 'latest');
+    if (release instanceof Error) {
+      return release;
+    }
+    let latest = release.tag_name; 
+
+    let checkForUpdates = vscode.workspace.getConfiguration()!.get<string>(config.cliUpdates);
+
+    if (!current) {
+      let text = `The Buf CLI could not be found on your PATH.\nWould you like to download and install buf ${latest}`;
+      if(await vscode.window.showInformationMessage(text, `Install buf ${latest}`)) {
+        return this.install(release);
+      }
+    } else if (semver.lt(current, latest) && checkForUpdates) {
+      let text = `A new version of the Buf CLI is available. Would you like to upgrade? ${current} -> ${latest}`;
+      let update = `Install buf ${latest}`;
+      let dontCheck = "Don't ask again";
+      
+      return vscode.window.showInformationMessage(text, update, dontCheck).then(async (choice) => {
+        switch (choice) {
+          case update:
+            let result = await this.install(release);
+            if (result instanceof Error) {
+              return result;
+            }
+
+            let text = `A new version of the Buf CLI (${latest}) has been installed.`;
+            if (await vscode.window.showInformationMessage(text, "Restart Language Server")) {
+              vscode.commands.executeCommand("buf.restartServer");
+            }
+            break;
+          case dontCheck:
+            vscode.workspace.getConfiguration()!.update(config.cliUpdates, false, vscode.ConfigurationTarget.Global);
+            break;
+        }
+        return;
+      });
+    }
+  }
+
+  private async install(release: github.Release): Promise<Result<void>> {
+    let platform = getPlatform();
+    if (platform instanceof Error) {
+      return platform;
+    }
+    let name = `buf-${platform}.tar.gz`;
```

Comment:
We need a special case for windows where we look for the `.exe` file instead of the tarball.